src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
             var builder = ArrayBuilder<Result>.GetInstance(n);
             for (int i = 0; i < n; i++)
             {
-                RefKind refKind = GetRefKind(refKindsOpt, i);
+                VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                builder.Add(_result);
+            }
+
+            _result = _invalidType;
+            return builder.ToImmutableAndFree();
+        }
+
+        private void VisitArgumentEvaluate(ImmutableArray<BoundExpression> arguments, ImmutableArray<RefKind> refKindsOpt, int i)
+        {
+            RefKind refKind = GetRefKind(refKindsOpt, i);
+            var argument = arguments[i];
+            if (refKind != RefKind.Out)
+            {
+                // PROTOTYPE(NullReferenceTypes): `ref` arguments should be treated as l-values
+                // for assignment. See `ref x3` in StaticNullChecking.PassingParameters_01.
+                VisitRvalue(argument);
+            }
+            else
+            {
+                VisitLvalue(argument);
+            }
+        }
+
+        /// <summary>
+        /// Visit all the arguments for the purpose of computing the exit state of the method,
+        /// given the annotations, but disabling warnings.
+        /// </summary>
+        private void VisitArgumentsEvaluateHonoringAnnotations(
+            ImmutableArray<BoundExpression> arguments,
+            ImmutableArray<RefKind> refKindsOpt,
+            ImmutableArray<ParameterSymbol> parameters,
+            ImmutableArray<int> argsToParamsOpt,
+            bool expanded)
+        {
+            Debug.Assert(!IsConditionalState);
+
+            ImmutableArray<(bool notNullWhenFalse, bool ensuresNotNull)> annotations = GetAnnotations(arguments.Length,
+               expanded, parameters, argsToParamsOpt);
+
+            if (annotations.IsDefault)
+            {
+                return;
+            }
+
+            bool saveDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+
+            Debug.Assert(annotations.Length == arguments.Length);
+
+            // Since we may be splitting states to represent the method returning true/false,
+            // we'll prepare all the slots ahead of time.
+            var slots = ArrayBuilder<int?>.GetInstance(arguments.Length);
+            for (int i = 0; i < arguments.Length; i++)
+            {
                 var argument = arguments[i];
-                if (refKind != RefKind.Out)
+                int slot = MakeSlot(argument);
+                if (slot <= 0)
+                {
+                    slots.Add(null);
```

Comment:
Consider using `slot` as is, for consistency with the rest of the class, rather than using `null` to represent no slot.